### DISCUSSION TITLE
CS0177 weird behavior

### DISCUSSION THREAD

**Discussion by levicki** [dotnet/roslyn#69057] [General] [answered]
Please consider the following code:
```
// Case 1 - CS0177 error
private bool GetSecureKeyDetail(UInt32 Timeout, out UInt32 FuncMask)
{
	IntPtr ResultPtr = IntPtr.Zero;

	bool Succeeded = false;

	try {
		Int32 hr = GetInfo(WFS_INF_PIN.SECUREKEY_DETAIL, IntPtr.Zero, Timeout, out ResultPtr);

		if (hr == WFS_ERR.SUCCESS) {
			WFSRESULT Result = Marshal.PtrToStructure<WFSRESULT>(ResultPtr);

			WFSPINSECUREKEYDETAIL SKDetail = Marshal.PtrToStructure<WFSPINSECUREKEYDETAIL>(Result.Buffer);

			WFSPINFUNCKEYDETAIL FKDetail = Marshal.PtrToStructure<WFSPINFUNCKEYDETAIL>(SKDetail.FuncKeyDetail);

			FuncMask = FKDetail.FuncMask;

			Succeeded = true;
		}
	} finally {
		if (!Succeeded) {
			FuncMask = 0;
		}
		if (ResultPtr != IntPtr.Zero) {
			FreeResult(ResultPtr);
		}
	}

	return Succeeded;
}
```
This code generates CS0177 error which I believe should not be generated because if I am not mistaken `FuncMask` will always be written.

On the other hand, if I modify the code to look like this:
```
// Case 2 - no CS0177 error
private bool GetSecureKeyDetail(UInt32 Timeout, out UInt32 FuncMask)
{
	IntPtr ResultPtr = IntPtr.Zero;

	bool Succeeded = false;

	try {
		Int32 hr = GetInfo(WFS_INF_PIN.SECUREKEY_DETAIL, IntPtr.Zero, Timeout, out ResultPtr);

		if (hr == WFS_ERR.SUCCESS) {
			WFSRESULT Result = Marshal.PtrToStructure<WFSRESULT>(ResultPtr);

			WFSPINSECUREKEYDETAIL SKDetail = Marshal.PtrToStructure<WFSPINSECUREKEYDETAIL>(Result.Buffer);

			WFSPINFUNCKEYDETAIL FKDetail = Marshal.PtrToStructure<WFSPINFUNCKEYDETAIL>(SKDetail.FuncKeyDetail);

			FuncMask = FKDetail.FuncMask;

			Succeeded = true;
		} else {
			FuncMask = 0;
		}
	} finally {
		if (ResultPtr != IntPtr.Zero) {
			FreeResult(ResultPtr);
		}
	}

	return Succeeded;
}
```
The CS0177 error goes away, but if I am not mistaken it should be generated because if `Marshal.PtrToStructure` throws then `FuncMask` won't be written at all.

Please share your thoughts on CS0177 correctness, should this be reported as an issue or not?

**Comment by CyrusNajmabadi** [comment]
This is by design. The language doesn't track the relationship between disparate variables. If you would like that, you'll need to start a discussion on csharplang (or add to an existing one).

**Reply by levicki** [reply]
@CyrusNajmabadi

I was wondering if someone can explain what is the purpose of `CS0177` error detection if it can't actually detect whether the `out` value is or isn't being writen? Why bother me with showing an error during compile at all if you can't show it when I make an actual mistake?

If it doesn't track relationships, then how can it claim with such certainty that the value isn't written? Shouldn't this be a warning rather than a compile error then? As in "maybe you should check if this is actually written", instead of forcing everyone to adopt code writting pattern which will lead to unnecessary duplicate writes to the same memory location in 99.99% of the cases?

It's again the same pattern of "feature X does Y" written on the tin, which when dissected turns into "But feature X does only a small subset of Y". At this point this MVP pattern has become worse than the labels on snake oil nutritional supplements.

**Reply by svick** [reply]
If this was an error instead of a warning, it would become quite common that you have an uninitialized variable, because of a bug in your code. Which you really don't want. And so C# errs on the side of safety and sometimes makes you assign the variable when it's not strictly necessary,

**Reply by CyrusNajmabadi** [reply]
> I was wondering if someone can explain what is the purpose of CS0177 error detection if it can't actually detect 

Actual error cases are detected. And many/most success cases are fine for the language/compiler. But as it is conservative analysis, there are false positives in some cases

**Reply by CyrusNajmabadi** [reply]
> If it doesn't track relationships, then how can it claim with such certainty that the value isn't written?

It doesn't claim with certainty.  It's the opposite.  It's not certain it's written, so it errors.  Only when it is *certain* is it ok.

> Shouldn't this be a warning rather than a compile error then? 

No.  Because then users could trivially run into uninitialized variables, which we think is bad.

> As in "maybe you should check if this is actually written", instead of forcing everyone to adopt code writting pattern which will lead to unnecessary duplicate writes to the same memory location in 99.99% of the cases?

You're welcome to suggest changes to the language. As i mentioned, that would go to dotnet/csharplang.  Roslyn is behaving as per the spec here.  If you want changes to the spec, they go there, not here. :)

**Reply by levicki** [reply]
> Actual error cases are detected.

And yet my examples show they are not. Two out of two wrong detections &mdash; 100% failure rate.

> It doesn't claim with certainty.

Yes it does &mdash; showing an error means it is certain that what I wrote is wrong and it wants me to correct it, while showing a warning would mean it isn't certain and it wants me to double-check the correctness myself.

> Only when it is certain is it ok.

And yet my code example 2 shows how its certainness is plain incorrect because variable can end up uninitialized.

> Because then users could trivially run into uninitialized variables, which we think is bad.

They can already trivially run into uninitialized variables as demonstrated above with case 2.

> You're welcome to suggest changes to the language.

Fat lot of good will that do when we can't even agree that the spec is broken.

**Reply by CyrusNajmabadi** [reply]
>  Two out of two wrong detections — 100% failure rate.

I don't know how you're computing this.  In one case it's a not a failure. We can't prove it's safe, so we report that.  

For your second case, this is working as intended.  In the case of an exception, the out variables don't need to be written (just like return values aren't provided).  Why?  Because there was an exception, so the caller won't complete on the call anyways to write into the out-vars they pass in, or execute any assignments of the call the make.  

> Yes it does — showing an error means it is certain that what I wrote is wrong and it wants me to correct it

No.  It means it hasn't proven it is safe.  When it can't prove it is safe, it is not legal as per the the rules of the language.  Again, if you want the language to change here you have to petition the language.

**Reply by CyrusNajmabadi** [reply]
> And yet my code example 2 shows how its certainness is plain incorrect because variable can end up uninitialized.

The variable is not initialized.  But the caller cannot observe that.  Why?  Because an exception was thrown.    In all cases where the code runs to an actual method exit successfullly (non-exceptionally) the value must be written.

> They can already trivially run into uninitialized variables as demonstrated above with case 2.

Your code does not demontrate it at all.  It demonstrates not writing in the case of an exception.  It does not demonstrate any code being able to *run into* the uninitialized variable.

> Fat lot of good will that do when we can't even agree that the spec is broken.

This attitude is not helpful.  Right now the language is spec'ed very specifically here, and the compiler is abiding by that specification.  If you want the compiler to do something else, you will need to petition for a language change.  That's how things work here.

**Reply by CyrusNajmabadi** [reply, marked as answer]
Simplified:

In the first case, there appears to be a non-exceptional codepath where the variable is not assigned. This is because the language doesn't track the relationship between variables, so we don't understand how to use the 'Succeeded' variable to inform things here. 

If you want that to be supported, you'll have to make a language changes request. 

In the second case, we can see that in all non-exceptional paths it is assigned.  And it exceptional paths it isn't(which is 100% ok). This is again working just as spec'ed in the language. There's no indication that there is anything even wrong in this case. This works very much as intended wrt exceptional vs non-exceptional flows.  You could of course request a change, but no argument has been put forth about why this behavior is undesirable for the second case.

**Reply by CyrusNajmabadi** [reply]
Intuition, for anyone interested (though I recommend reading the actual spec). All out variables need to have been written by the time any 'end point' of the method is reached. You can reach those end points by hitting a return statement, or naturally hitting the end of the method block itself.   Where naturally is normal (non-exceptional) code flow. 

In the second example, the only way to naturally hit the actual end points of the method always go through codepaths that assign the out var. The only way to not hit it is with an exception.  But that's completely fine. A more trivial way if seeing that is just:

```c#
void M(out int x) => throw new Not implemented exception();
```

This is entirely fine as the caller of M would never know the difference. Either you assign 'x' and return successfully, and this that see that value. Or you throw, and don't assign, but then the caller never can tell because the exception continues to bubble up.

**Reply by CyrusNajmabadi** [reply]
Precise spec rules. 

> Every output parameter of a function member or anonymous function shall be definitely assigned ([§9.4](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/variables.md#94-definite-assignment)) before the function member or anonymous function returns normally.

Definite assignment: https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/variables.md#94-definite-assignment

> All output parameters of a function member shall be definitely assigned at each location where the function member returns (through a return statement or through execution reaching the end of the function member body).

**Reply by levicki** [reply]
> If you want that to be supported, you'll have to make a language changes request.

I would definitely want that to be supported because to me it makes sense -- compiler should not claim an error because it can't "see" the assignment.

As for the second case I stand corrected &mdash; this is actually why I started this topic as a discussion because I suspected I might be overlooking something. I will now mark your comment which I expected to get instead of "works to spec, closed" as an answer. You have my genuine thanks for engaging further despite my attitude.

**Reply by CyrusNajmabadi** [reply]
> I would definitely want that to be supported because to me it makes sense -- compiler should not claim an error because it can't "see" the assignment.

Then please either open a discussion or continue a discussion at dotnet/csharplang.  That's a request to change the language rules. If that change is approved, roslyn will be updated to follow the spec. Thanks.

**Reply by CyrusNajmabadi** [reply]
> As for the second case I stand corrected — this is actually why I started this topic as a discussion because I suspected I might be overlooking something. 

Definitely. Note: the responses were very much simply that you are posting questions about how the language is working in the *compiler* repo. For context, the language defines how things are supposed to work and what is legal or not. Roslyn is then just an implementation of the rules. 

So if you're thinking that the compiler is actually incorrect, then this is the right place to bring things up (like the compiler doesn't match the language). If, however, you don't like how the *language* works, then this is not the right place. 

Ideally, posts get placed in the best location as soon as possible to keep things on topic. 

>. You have my genuine thanks for engaging further despite my attitude.

You're welcome. :-)

**Reply by CyrusNajmabadi** [reply]
> compiler should not claim an error because it can't "see" the assignment.

At the end of the day, the compiler can only prove what it can prove. What if the relationship between the variables involves arithmetic relationships like in the collatz conjecture?

You have convinced yourself this is safe. If you want the language to think it is safe as well, you'll have to come up with a set of rules the language and compiler can implement to have this understanding.  Def check out the spec here and how it works. If you think your idea is possible (which I definitely think it is) the next steps would be a proposal that lays out the rules with reasonable precision (or outlines them with enough clarity that they could clearly be added to the definite-assignment specification). 

We would also need to check carefully that such an algorithm could clearly be added to the compiler without causing serious problems. For example, if it required tons of memory tracking the relationships between all variables through all codepaths, or required a lot of cpu to do this work.  


For context, nullable analysis does a ton of work to provide a decent null-tracking heuristic. And it caused major issues with both memory and cpu (something we're still dealing with today). For nullable it was worth it given the massive value in mitigating null-ref issues. If the same perf issues were present here, it would then be very hard to justify given it's easy to just disallow the code and have the user write in a way that is easier to prove. 

That said, if you find an awesome algorithm with great memory and cpu costs, I'd def be interested in driving this along.

**Reply by levicki** [reply]
> So if you're thinking that the compiler is actually incorrect, then this is the right place to bring things up (like the compiler doesn't match the language). If, however, you don't like how the language works, then this is not the right place.

But that's the point &mdash; I wasn't sure if the case 1 was a false positive (compiler error), or a language spec issue. That's why I asked here first and didn't even open an issue, but a discussion expecting to learn something. However, regardless of your intent the initial answer came across as dismissive hence my attitude.

> At the end of the day, the compiler can only prove what it can prove.

And this is where we disagree on error .vs. warning &mdash; if you can't prove my code has an error then don't emit an error.

I come from C/C++ background, and this rather contrived (and hopefully a bit humorous) example shows what kind of behavior I expect from compilers:
```
#include <random>
#include <stdexcept>
#include <stdio.h>
#include <stdlib.h>
#include <Windows.h>

static std::mt19937 rng;
static std::uniform_int_distribution<int> dice(1, 6);

int meaningoflife()
{
	int roll = dice(rng);

	wprintf(L"Dice : %d\n", roll);

	if (roll == 1) {
		throw std::runtime_error("oopsie!");
	} else {
		return 42;
	}
}

bool somefunc()
{
	int answer;
	bool succeeded;

	try {
		answer = meaningoflife();
		succeeded = true;
	} catch (...) {
		answer = 2; // FILE NOT FOUND
		succeeded = false;
	}
	
	return (answer == 42) && succeeded;
}

int wmain(int argc, wchar_t *argv[])
{
	UNREFERENCED_PARAMETER(argc);
	UNREFERENCED_PARAMETER(argv);

	rng.seed(GetTickCount());

	if (somefunc()) {
		wprintf(L"SUCCESS\n");
	} else {
		wprintf(L"FAILURE\n");
	}
	
	return 0;
}
```
Compiling this at `/W4` warning level produces no warnings at all. Why? Because the code, as silly as it may be, is correct.

But comment out `succeeded = false` in `catch` block and you will get:
```
warning C4701: potentially uninitialized local variable 'succeeded' used
```
Even though the compiler can be reasonably sure that `succeeded` is not initialized before use, it still doesn't throw an error &mdash; it just warns me that I should check my code.

You can say "aha, but in C/C++ external code could theoretically write to the memory location of `succeeded` variable via some aliased pointer so the compiler can't prove `succeeded` is uninitialized hence a warning" and I can say "C# compiler can't prove I did not write into `FuncMask` even if it's right there in front of its nose, but it still confidently throws an error".

If C++ compiler was as anally retentive as C# compiler, I would have to write `somefunc` like this:
```
bool somefunc()
{
	int answer = 2; // FILE NOT FOUND
	bool succeeded = false;

	try {
		answer = meaningoflife();
		succeeded = true;
	} catch (...) {
		// ?
	}
	
	return (answer == 42) && succeeded;
}
```
The resulting native code would look like this:
```
somefunc        proc near

succeeded       = byte ptr  8
answer          = dword ptr  10h

; __unwind { // sub_140002F68
                sub     rsp, 28h
                mov     [rsp+28h+answer], 2 ; answer = 2
                mov     [rsp+28h+succeeded], 0 ; succeeded = false
                call    meaningoflife
                mov     [rsp+28h+answer], eax ; return value (if any) from meaningoflife
                mov     cl, 1
                mov     [rsp+28h+succeeded], cl ; succeeded = true
                jmp     short loc_14000167B
                mov     eax, [rsp+28h+answer]
                movzx   ecx, [rsp+28h+succeeded]
loc_14000167B:
                cmp     eax, 2Ah ; '*'
                jnz     short loc_14000168B
                test    cl, cl
                jz      short loc_14000168B
                mov     al, 1
                add     rsp, 28h
                retn
loc_14000168B:
                xor     al, al
                add     rsp, 28h
                retn
; } // starts at 140001650
somefunc        endp
```
And if the code is written so that variables are initialized in `catch` block, you get this:
```
somefunc        proc near               ; CODE XREF: main+18\u2193p
; __unwind { // sub_140002E08
                sub     rsp, 28h
                call    meaningoflife
                nop
                cmp     eax, 2Ah ; '*'
                jnz     short loc_14000154B
                mov     al, 1
                jmp     short loc_14000154D
loc_14000154B:
                xor     al, al
loc_14000154D:
                add     rsp, 28h
                retn
; } // starts at 140001538
somefunc        endp
```
Now I don't know how smart C# compiler is to eliminate and optimize redundant code, but as you can see when you write explicit initialization at the start of the method (which is something CS0177 is practically enforcing), you get a lot of redundant memory writes. In this exaggerated case we know that `meaningoflife` will throw 1 out of 6 times on average, so ~83% of calls to it will succeed and in those 83% of calls you will write to memory twice.

So even if CS0177 might look like it is encouraging good coding style (upfront initialization instead of doing it in each code path), it most likely results in worse performing and larger code.

**Reply by CyrusNajmabadi** [reply]
> And this is where we disagree on error .vs. warning — if you can't prove my code has an error then don't emit an error.

This is the position of hte language and the design team here.  The language is opinionated in this area.  If it can't prove the code is correct, it does not allow it.  This has been the behavior for literally 25 years now at this point.  It was a core design decision for the language, and nothing has changed wrt our thinking on that in this area of the design.

Any change here would have to come by making (or continuing) a discussion over at csharplang where the design team can discuss the topic and the community and ecosystem can weigh in here on what the right direction forward is for the language.

We are not going to change it in roslyn because one person filed a request for the compiler to behave more like C++. :)

**Reply by CyrusNajmabadi** [reply]
> I come from C/C++ background

So do many of us.  And C# very intentionally is not designed to work like C/C++, especially in ways that we view as problematic.  We do not look to it as a good example on how a language should work. :)

This is also, for example, why we do not have automatic fallthrough in switch statements.  Or other C/C++ allowances we view as problematic.  Our goal is not to be a continuation of C++.  The goal is to be a language that does what we think is appropriate, using a syntax derived from the C lineage of languages :)

**Reply by CyrusNajmabadi** [reply]
> hence my attitude.

> "Fat lot of good will that do when we can't even agree that the spec is broken."

The posts you were responding to were literally welcoming you to take your feedback to the language.  I get that we've made decisions you don't agree with.  However, your response in these cases is often immediately disruptive and not conducive to constructive discussion.  Posts welcoming you to the right venue for your feedback are made in the effort to actually streamline things, so that the right people are involved with assessing and responding to your feedback.

**Reply by CyrusNajmabadi** [reply]
> and I can say "C# compiler can't prove I did not write into FuncMask even if it's right there in front of its nose, but it still confidently throws an error".

Yes.  That's the point.  If we cannot prove that it is safe, we presume it is unsafe.  That's literally the design here, and it is very intentional.  As per above, you are welcome to disagree with that design, and think it is too burdensome.

However, such feedback *here* won't accomplish anything *as has been repeated numerous times*.  This is the repo for the *compiler*, and the compiler is correctly implementing the language specification.  If you want the language specification to change to be more like how you view htings *you need to open, or continue, a discussion over at dotnet/csharplang*.  I've looked now, and it's like a dozen times this has been stated.  At this point, i'm not sure what else can make this clearer :-/

**Reply by CyrusNajmabadi** [reply]
> So even if CS0177 might look like it is encouraging good coding style (upfront initialization instead of doing it in each code path), it most likely results in worse performing and larger code.

Once again, feel free to suggest a change to the language over at dotnet/csharplang.  Profiling data would be valuable information when making your case.  The compiler (*this repo*) is not going to change until that happens. :)

**Comment by CyrusNajmabadi** [comment]
Intentionally locking.  This question has been asked and answered.  The latest responses are all things that should go to dotnet/csharplang instead.  Locking this so that that happens and the requests for language changing doing continue happening here.
